### PR TITLE
(CDAP-3716) Refactor code that causes deadlock

### PR DESCRIPTION
**Pull request by chtyim**
- Remove unnecessary synchronize method
- Always update the in memory runtimeInfos from the controller callback thread
- Remove bunch of unused code

### CODE REVIEW

**Comment by chtyim** [comment]
Build passed: http://builds.cask.co/browse/CDAP-RBT455

**Comment by chtyim** [comment]
Verified on cluster for the latest changes.

File: cdap-app-fabric/src/main/java/co/cask/cdap/app/runtime/AbstractProgramRuntimeService.java [hunk 1]
```diff
@@ -287,14 +281,20 @@ protected void updateRuntimeInfo(ProgramType type, RunId runId, RuntimeInfo runt
     lock.lock();
```
Comment (gokulavasan) [inline cdap-app-fabric/src/main/java/co/cask/cdap/app/runtime/AbstractProgramRuntimeService.java]: Do we need to place a writeLock here? What write operation are we performing?

Comment (chtyim) [inline cdap-app-fabric/src/main/java/co/cask/cdap/app/runtime/AbstractProgramRuntimeService.java]: right, a read lock should be enough.

**Comment by gokulavasan** [comment]
LGTM :+1:

**Pull request merged**